content/blog/2019-09-25-interpreter-vector-support/index.md

```diff
@@ -0,0 +1,134 @@
+
+It's awkward that the interpreter supports vector instructions, but doesn't actually use a native vector assembly instruction to perform the computation. We expect that the performance of the interpreted vector instructions will be poor. To explore this hypothesis, we create a version of an interpreted vector add instruction using three methods: 1) Typescript, 2) serial c++, and 3) vectorized c++. We run each test for 10,000 iterations and average the execution time over five runs. We assumed 10,000 iterations was enough time for the Typescript/Javascript JIT to warmup. The average time per loop iteration for each implementation is given below.
+
+|     Method    | Time per iteration (ns) |
+| ------------- | ------------- |
+| Typescript  | 317  |
+| Serial C++  | 16  |
+| Vector C++  | 9  |
+
+There is a fair benefit to using native vector instruction from C (35x) speedup over the Typescript version. A large portion of this benefit comes from just using C in the first place. The serial C version achieves 20x better performance while the vector version further improves the performance by a more modest 1.8x.
+
+### C++ binding for Typescript
+
+In order to utilize vector intrinsics, we need to call the C++ implementations in the Bril interpreter. We use [nbind](https://github.com/charto/nbind) to allow Typescript to execute binaries generated from C++. These sorts of calls will add potentially significant overhead to the execution. We quantify this overhead to see if it is practical in the intepreter. Note that each time we run a single vector instruction we must make a call to the binding. We run a vector add program with various iterations. Each iteration does two vector loads, a vector add, and a vector store along with instructions to facilitate the iteration. We run five configurations (128, 1024, 2048, 4196, 8192) and average the execution time over five runs. We compare the execution time with and without calls to the binding (literally just comment the line out). Note that the calls include passing arguments to the C++ binary, which incurs some additional overhead. On average there is a 10% overhead in the program due to the binding call. This overhead is expected to be offset from the substantial speedups offered by the C++ implementation.
+
+## Evaluation
+
+### Correctness
+
+We write multiple test programs to verify that the memory and vector instructions functioned as expected. [Turnt](https://github.com/cucapra/turnt) is used to test the expected output of the program from `print` instructions. The vector programs could not be verified with Turnt, however, because it needs to be executed in the interpreter directory to find the location of the C++ binaries. These programs were verified by manually inspecting the output. We test a simple store and then load, multiple stores and then multiple loads, and vvadd with both the Typescript implementation and the C++ vector implementation. All functioned as specified in this document.
```

Comment:
I think you should be able to make Turnt work—you could, like, `cd /path/to/interp ; ...` to hack it?